### PR TITLE
Add dev-python/jedi-language-server

### PR DESCRIPTION
**Pull request by mattst88**
I've been using [dev-python/python-language-server](https://github.com/palantir/python-language-server) (via [coc.nvim](https://github.com/neoclide/coc.nvim)) but it seems to be mostly unmaintained upstream, so I looked for an alternative and found [jedi-language-server](https://github.com/pappasam/jedi-language-server).

This PR adds it and some dependent packages:

* [dev-python/typeguard](https://github.com/agronholm/typeguard)
* [dev-python/pygls](https://github.com/openlawlibrary/pygls/)
* [dev-python/docstring-to-markdown](https://github.com/krassowski/docstring-to-markdown)
* [dev-python/jedi-language-server](https://github.com/pappasam/jedi-language-server)

I'd be happy to see @gentoo/python and/or @AndrewAmmerlaan as additional maintainers. I'd really appreciate some review as well—I'm not often packaging `dev-python` things.

Except for a single test in `typeguard`, all the packages' test suites run and pass.

### CODE REVIEW

**Comment by mattst88** [comment]
> New issues caused by PR:
> https://qa-reports.gentoo.org/output/gentoo-ci/245a62cb2f/output.html#dev-python/typeguard

This is kind of a hack because the package's tests fail without the package previously installed.

File: dev-python/jedi-language-server/jedi-language-server-0.30.2.ebuild [hunk 1]
```diff
@@ -0,0 +1,41 @@
+# Copyright 1999-2021 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=7
+
+PYTHON_COMPAT=( python3_{7..9} )
+
+DISTUTILS_USE_SETUPTOOLS=pyproject.toml
```
Comment (mattst88) [inline dev-python/jedi-language-server/jedi-language-server-0.30.2.ebuild:8]: Ugh, I just remembered that this is broken.

* The PyPI tarball doesn't include tests but builds/installs successfully with `DISTUTILS_USE_SETUPTOOLS=rdepend`.
* The GitHub tarball includes tests but doesn't include `setup.py` (so `DISTUTILS_USE_SETUPTOOLS=rdepend` doesn't work) but `DISTUTILS_USE_SETUPTOOLS=pyproject.toml` seems to fail to install most of the package. Maybe a bug in `dev-python/pyproject2setuppy` @mgorny?

File: dev-python/docstring-to-markdown/docstring-to-markdown-0.9.ebuild [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+# PyPI tarball doesn't include tests
+SRC_URI="https://github.com/krassowski/${PN}/archive/v${PV}.tar.gz -> ${P}.tar.gz"
+
+LICENSE="LGPL-2.1"
+SLOT="0"
+KEYWORDS="~amd64"
+
+DEPEND="
+	test? (
+		dev-python/pytest-cov[${PYTHON_USEDEP}]
```
Comment (AndrewAmmerlaan) [inline dev-python/docstring-to-markdown/docstring-to-markdown-0.9.ebuild:21]: `pytest-cov` is an undesirable test dependency according to: https://dev.gentoo.org/~mgorny/python-guide/distutils.html#enabling-tests

Usually it is possible to remove this dependency by editing the pytest.ini or setup.cfg file

File: dev-python/pygls/pygls-0.10.2.ebuild [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+DESCRIPTION="A pythonic generic language server"
+HOMEPAGE="https://github.com/openlawlibrary/pygls"
+SRC_URI="mirror://pypi/${PN:0:1}/${PN}/${P}.tar.gz"
+
+LICENSE="Apache-2.0"
+SLOT="0"
+KEYWORDS="~amd64"
+IUSE="test"
+RESTRICT="!test? ( test )"
```
Comment (AndrewAmmerlaan) [inline dev-python/pygls/pygls-0.10.2.ebuild:18]: `distutils_enable_tests` already automatically adds the IUSE and RESTRICT lines for you

File: dev-python/typeguard/typeguard-2.12.0.ebuild [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+DESCRIPTION="Run-time type checker for Python"
+HOMEPAGE="https://github.com/agronholm/typeguard"
+SRC_URI="mirror://pypi/${PN:0:1}/${PN}/${P}.tar.gz"
+
+LICENSE="MIT"
+SLOT="0"
+KEYWORDS="~amd64"
+IUSE="test"
+RESTRICT="!test? ( test )"
```
Comment (AndrewAmmerlaan) [inline dev-python/typeguard/typeguard-2.12.0.ebuild:18]: Same here

**Comment by AndrewAmmerlaan** [comment]
> I've been using [dev-python/python-language-server](https://github.com/palantir/python-language-server) (via [coc.nvim](https://github.com/neoclide/coc.nvim)) but it seems to be mostly unmaintained upstream, so I looked for an alternative and found [jedi-language-server](https://github.com/pappasam/jedi-language-server).

There is also this: https://github.com/python-ls/python-ls which is where most of the development for pyls seems to have moved to. This is where I got the py3.9 patches for pyls.

File: dev-python/docstring-to-markdown/docstring-to-markdown-0.9.ebuild (hunk 2 above)
Comment (mattst88) [reply dev-python/docstring-to-markdown/docstring-to-markdown-0.9.ebuild]: Thanks, fixed with some `sed`.

File: dev-python/pygls/pygls-0.10.2.ebuild (hunk 3 above)
Comment (mattst88) [reply dev-python/pygls/pygls-0.10.2.ebuild]: Thanks, fixed.

File: dev-python/typeguard/typeguard-2.12.0.ebuild (hunk 4 above)
Comment (mattst88) [reply dev-python/typeguard/typeguard-2.12.0.ebuild]: Thanks, fixed.

File: dev-python/pygls/pygls-0.10.2.ebuild [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+		${RDEPEND}
+		dev-python/mock[${PYTHON_USEDEP}]
+		dev-python/pytest-asyncio[${PYTHON_USEDEP}]
+	)
+"
+
+distutils_enable_tests pytest
+
+python_prepare_all() {
+	sed -i -e "s/find_packages(exclude=\['tests'\])/find_packages(exclude=['tests*'])/" setup.py || die
```
Comment (mattst88) [inline dev-python/pygls/pygls-0.10.2.ebuild:35]: FWIW, I opened an upstream PR to fix this here: https://github.com/openlawlibrary/pygls/pull/169

File: dev-python/jedi-language-server/jedi-language-server-0.30.2.ebuild (hunk 1 above)
Comment (mattst88) [reply dev-python/jedi-language-server/jedi-language-server-0.30.2.ebuild:8]: This is now fixed per https://github.com/mgorny/pyproject2setuppy/issues/14

**Comment by AndrewAmmerlaan** [comment]
> > I've been using [dev-python/python-language-server](https://github.com/palantir/python-language-server) (via [coc.nvim](https://github.com/neoclide/coc.nvim)) but it seems to be mostly unmaintained upstream, so I looked for an alternative and found [jedi-language-server](https://github.com/pappasam/jedi-language-server).
> 
> There is also this: https://github.com/python-ls/python-ls which is where most of the development for pyls seems to have moved to. This is where I got the py3.9 patches for pyls.
> 
> See Also: [python-lsp/python-lsp-server#4](https://github.com/python-lsp/python-lsp-server/issues/4) and [palantir/python-language-server#901 (comment)](https://github.com/palantir/python-language-server/pull/901#issuecomment-771180214)

I added `dev-python/python-lsp-server` yesterday, which is the fork where most/all development of pyls seems to have moved to. I hope this will resolve most of the issues we've been having with pyls (outdated deps and all that mess). We can still add jedi-language-server as well if it is of added value.

**Comment by AndrewAmmerlaan** [comment]
> I added `dev-python/python-lsp-server` yesterday, which is the fork where most/all development of pyls seems to have moved to. I hope this will resolve most of the issues we've been having with pyls (outdated deps and all that mess). We can still add jedi-language-server as well if it is of added value.

Closing this in favour of `python-lsp-server` which seems to be doing a better job at keeping things up to date now. Please re-open if we should still add `jedi-language-server`

**Pull request closed without merging**